Add optimizing AT_LIGHT_PASS builtin to canvas shaders (2.1)

RandomShaper opened the pull request:
This one allows for complex shaders paired with a simple lighting shader to skip code that would otherwise be pointlessly (and wastefully) run during the light pass.

You can use `if (AT_LIGHT_PASS)`, negated or not, and that will be converted to a preprocessed #if when the shader is compiled.

Depending on your game (number of items and lights), this can be a *significant* performance gain, or at least avoids relying on the driver's optimizing abilities.

~~I don't think this is relevant for _master_.~~ @Zireael07 is right; now I'm porting it.

Zireael07 commented:
> I don't think this is relevant for master.

Why would it not be? Skipping code during light pass should be useful for any renderer....

RandomShaper commented:
My plan was having a look to _master_ later, but I had the idea things are done so differently in 3.0 that something like this is not needed or can be done by builtin means.

But maybe you're right.

RandomShaper committed:
Add optimizing AT_LIGHT_PASS builtin to canvas shaders

This one allows for complex shaders paired with a simple lighting shader to skip code that would otherwise be pointlessly (and wastefully) run during the light pass.
You can use `if (AT_LIGHT_PASS) , negated or not, and that will be converted to a preprocessed #if when the shader is compiled.
Depending on your game (number of items and lights), this can be a *significant* performance gain, or at least avoids relying on the driver's optimizing abilities.

akien-mga commented:
@RandomShaper Is this one ready to merge as is or do you want to edit it based on the discussions on the master PR?

RandomShaper commented:
I did the part about renaming/less-uglyness; the conversion to preprocessed `#if` is still in this version as maybe meaningful for some hardware.

So ready for merging!
